File: src/main/java/carelog/carelog/user/app/UserServiceImpl.java
package carelog.carelog.user.app;

import carelog.carelog.auth.app.CustomUserDetails;
import carelog.carelog.common.web.exception.*;
import carelog.carelog.user.domain.*;
import carelog.carelog.user.web.dto.*;
import lombok.*;
import org.springframework.security.crypto.password.*;
import org.springframework.stereotype.*;
import org.springframework.transaction.annotation.*;

import java.util.List;
import java.util.UUID;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class UserServiceImpl implements UserService {

    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;

    private User findUserEntityByPublicId(UUID publicId) {
        return userRepository.findByPublicId(publicId)
                .orElseThrow(() -> new CustomException(ExceptionStatus.USER_NOT_FOUND));
    }

    @Override
    @Transactional
    public UserResponse createManager(ManagerCreateRequest request) {
        if (userRepository.existsByUserId(request.userId())) {
            throw new CustomException(ExceptionStatus.DUPLICATE_USER_ID);
        }
        if (userRepository.existsByEmail(request.email())) {
            throw new CustomException(ExceptionStatus.DUPLICATE_EMAIL);
        }

        User newUser = User.builder()
                .userId(request.userId())
                .password(passwordEncoder.encode(request.password()))
                .email(request.email())
                .name(request.name())
                .role(UserRole.MANAGER)
                .managerType(request.managerType())
                .phoneEncrypted(request.phoneEncrypted())
                .addressEncrypted(request.addressEncrypted())
                .build();

        newUser.assignOrganization(UUID.randomUUID());
        User savedUser = userRepository.save(newUser);
        return UserResponse.from(savedUser);
    }

    @Override
    @Transactional
    public UserResponse createCustomer(CustomerCreateRequest request, CustomUserDetails userDetails) {
        User newUser = User.builder()
                .name(request.name())
                .role(UserRole.CUSTOMER)
                .build();

        newUser.assignOrganization(userDetails.getOrganizationId());
        User savedUser = userRepository.save(newUser);
        return UserResponse.from(savedUser);
    }

    @Override
    public UserResponse findUserByUserId(String userId) {
        User user = userRepository.findByUserId(userId)
                .orElseThrow(() -> new CustomException(ExceptionStatus.USER_NOT_FOUND));
        return UserResponse.from(user);
    }

    @Override
    public UserResponse findUserByEmail(String email) {
        User user = userRepository.findByEmail(email)
                .orElseThrow(() -> new CustomException(ExceptionStatus.USER_NOT_FOUND));
        return UserResponse.from(user);
    }

    @Transactional
    @Override
    public UserResponse updateUser(UUID publicId, UserUpdateRequest request) {
        User user = findUserEntityByPublicId(publicId);

        if (request.password() != null && !request.password().isBlank()) {
            user.updatePassword(passwordEncoder.encode(request.password()));
        }
        if (request.phoneEncrypted() != null) {
            user.updatePhoneEncrypted(request.phoneEncrypted());
        }
        if (request.addressEncrypted() != null) {
            user.updateAddressEncrypted(request.addressEncrypted());
        }

        return UserResponse.from(user);
    }

    @Override
    @Transactional
    public void deleteUser(UUID publicId) {
        User user = findUserEntityByPublicId(publicId);
        userRepository.delete(user);
    }

    @Override
    public List<UserResponse> findAllCustomers(String name) {
        List<User> customers = (name != null && !name.isBlank()) ?
                userRepository.findAllByRoleAndNameContaining(UserRole.CUSTOMER, name)
                : userRepository.findAllByRole(UserRole.CUSTOMER);

        return customers.stream()
                .map(UserResponse::from)
                .toList();
    }
}

File: src/main/java/carelog/carelog/user/app/UserServiceImpl.kt
package carelog.carelog.user.app

import carelog.carelog.common.web.exception.CustomException
import carelog.carelog.common.web.exception.ExceptionStatus
import carelog.carelog.user.domain.User
import carelog.carelog.user.domain.UserRepository
import carelog.carelog.user.domain.UserRole
import carelog.carelog.user.web.dto.UserCreateRequest
import carelog.carelog.user.web.dto.UserResponse
import carelog.carelog.user.web.dto.UserUpdateRequest
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional


@Service
@Transactional(readOnly = true)
class UserServiceImpl(
    private val userRepository: UserRepository,
    private val passwordEncoder: PasswordEncoder
): UserService {


    private fun findUserEntityById(id: Long): User =
         userRepository.findById(id)
                .orElseThrow { CustomException(ExceptionStatus.USER_NOT_FOUND)}

    @Transactional
    override fun createUser(request: UserCreateRequest): UserResponse {
        if (userRepository.existsByUserId(request.userId)) {
            throw CustomException(ExceptionStatus.DUPLICATE_USER_ID)
        }
        if (userRepository.existsByEmail(request.email)) {
            throw CustomException(ExceptionStatus.DUPLICATE_EMAIL)
        }

        val newUser = User(
            userId = request.userId,
            password = passwordEncoder.encode(request.password),
            email = request.email,
            name = request.name,
            role = UserRole.MANAGER,
            phoneEncrypted = request.phoneEncrypted,
            addressEncrypted = request.addressEncrypted
        )

        val savedUser: User = userRepository.save(newUser)
        return UserResponse.from(savedUser)
    }

    override fun findUserById(id: Long): UserResponse =
          UserResponse.from(findUserEntityById(id))


    override fun findUserByUserId(userId: String): UserResponse =
        userRepository.findByUserId(userId)
            .map{ UserResponse.from(it) }
            .orElseThrow { CustomException(ExceptionStatus.USER_NOT_FOUND)}

    override fun findUserByEmail(email: String): UserResponse =
        userRepository.findByEmail(email)
            .map { UserResponse.from(it) }
            .orElseThrow{ CustomException(ExceptionStatus.USER_NOT_FOUND) }

    @Transactional
    override fun  updateUser(id: Long, request: UserUpdateRequest): UserResponse {
        val user = findUserEntityById(id)

        request.password?.takeIf { it.isNotBlank() }?.let { user.updatePassword(it) }
        request.phoneEncrypted?.let { user.updatePhoneEncrypted(it) }
        request.addressEncrypted?.let { user.updateAddressEncrypted(it) }

        return UserResponse.from(user)
    }

    @Transactional
    override fun deleteUser(id: Long) {
        val user = findUserEntityById(id)
        userRepository.delete(user)
    }
}
